elements/lisk-utils/src/math/q.ts

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable no-bitwise */
+/*
+ * Copyright © 2022 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { MAX_FRAC, ONE, Q_OPERATION, TWO } from './constants';
+
+const numberToQ = (base: bigint, val: number): bigint => {
+	const denominator = TWO ** base;
+	if (Number.isInteger(val)) {
+		return BigInt(val) * denominator;
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		BigInt(intStr) * denominator +
+		(BigInt(fractionalStr) * denominator) / BigInt(10) ** BigInt(fractionalStr.length)
+	);
+};
+
+export const q = (val: number | bigint | Buffer, base: number | bigint): Q =>
+	Q.fromValue(val, base);
+
+export class Q {
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._base = BigInt(base);
+		if (this._base > MAX_FRAC) {
+			throw new Error('Base exceeds max fraction allowed.');
+		}
+		this._val = val;
+		this._base = BigInt(base);
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
+		if (val instanceof Q) {
+			return new Q(val._val, val._base);
+		}
+		const bigintBase = BigInt(base);
+		if (typeof val === 'number') {
+			return new Q(numberToQ(bigintBase, val), bigintBase);
+		}
+		if (typeof val === 'bigint') {
+			return new Q(val * TWO ** bigintBase, bigintBase);
+		}
+		return new Q(BigInt(`0x${val.toString('hex')}`), bigintBase);
+	}
+
+	public add(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Base of input does not match.');
```

Comment:
Instead of copying the same code in every public method, I think it is cleaner to handle that recurring task using a private method, e.g. `checkBase()`